### PR TITLE
Improve initialization functions

### PR DESCRIPTION
**Pull request by mariaschuld**
*Note: This PR forks from PR #409, which should be reviewed first.*

**Context:** Another PR in the templates refactor series. Resolves Issue #383, by changing the signature of initialization functions, adding more functions and improving the docs.

**Description of the Change:** 

* Every template now has initialization functions ``_uniform`` and ``_normal`` for every parameter array it takes. For example, ``Interferometer`` has three initializations for its three inputs ``theta``, ``phi`` ``varphi``.   
* Added ``_all`` initializations for ``Interferometer`` and ``CVNeuralNetLayers`` to initialize all parameters in one go.
* Added new functions to introduction in docs, and gave code examples.

**Benefits:** Hopefully resolves confusion around usage, improves explanations for the user, sets a clear standard of how new templates are treated.

**Possible Drawbacks:** Adds quite a lot of functions to the ``init`` module, with a lot of code repetition. The justification is that these functions are for convenience, and should be as simple as possible (which was why we previously decided for ``_normal`` and ``_uniform`` versions, rather than keyword arguments).

**Related GitHub Issues:** #383

### CODE REVIEW

File: doc/introduction/templates.rst [hunk 1]
```diff
@@ -159,10 +231,23 @@ randomly initialized arrays for the trainable parameters.
...
+    ~pennylane.init.cvqnn_layers_r_uniform
+    ~pennylane.init.cvqnn_layers_r_normal
+    ~pennylane.init.cvqnn_layers_phi_r_uniform
+    ~pennylane.init.cvqnn_layers_phi_r_normal
+    ~pennylane.init.cvqnn_layers_a_uniform
+    ~pennylane.init.cvqnn_layers_a_normal
+    ~pennylane.init.cvqnn_layers_phi_a_uniform
+    ~pennylane.init.cvqnn_layers_phi_a_normal
+    ~pennylane.init.cvqnn_layers_phi_a_uniform
+    ~pennylane.init.cvqnn_layers_phi_a_normal
```
Comment (josh146) [inline doc/introduction/templates.rst]: 😆

File: pennylane/__init__.py [hunk 2]
```diff
@@ -28,6 +28,7 @@
 import pennylane.init
 import pennylane.templates.layers
 import pennylane.templates.embeddings
+import pennylane.templates.subroutines
```
Comment (josh146) [inline pennylane/__init__.py]: Will all layers, embeddings, and subroutines be imported here, so that you can do `qml.templates.StronglyEntangling` etc?

File: pennylane/init.py [hunk 3]
```diff
@@ -33,22 +33,22 @@ def strong_ent_layers_uniform(n_layers, n_wires, low=0, high=2 * pi, seed=None):
...
     Keyword Args:
-        low (float): minimum value of non-angle gate parameters
-        high (float): maximum value of non-angle gate parameters
+        low (float): minimum value of uniform distribution
+        high (float): maximum value of uniform distribution
         seed (int): seed used in sampling the parameters, makes function call deterministic
 
     Returns:
-        list of parameter array
+        parameter array
```
Comment (josh146) [inline pennylane/init.py]:
```suggestion
        array: parameter array
```

File: pennylane/init.py [hunk 4]
```diff
@@ -144,7 +92,7 @@ def random_layers_uniform(n_layers, n_wires, n_rots=None, low=0, high=2 * pi, se
...
```
Comment (josh146) [inline pennylane/init.py]:
```suggestion
        array: parameter array
```

File: pennylane/init.py [hunk 5]
```diff
@@ -183,273 +131,586 @@ def random_layers_normal(n_layers, n_wires, n_rots=None, mean=0, std=0.1, seed=N
...
+    r = cvqnn_layers_r_normal(**kwargs)
+    phi_r = cvqnn_layers_phi_r_uniform(**kwargs)
+    theta_2 = cvqnn_layers_theta_uniform(**kwargs)
+    phi_2 = cvqnn_layers_phi_uniform(**kwargs)
+    varphi_2 = cvqnn_layers_varphi_uniform(**kwargs)
+    a = cvqnn_layers_a_normal(**kwargs)
+    phi_a = cvqnn_layers_phi_a_uniform(**kwargs)
+    k = cvqnn_layers_kappa_normal(**kwargs)
 
+    return [theta_1, phi_1, varphi_1, r, phi_r, theta_2, phi_2, varphi_2, a, phi_a, k]
```
Comment (josh146) [inline pennylane/init.py]: Nice!

File: pennylane/init.py [hunk 6]
```diff
@@ -183,273 +131,586 @@ def random_layers_normal(n_layers, n_wires, n_rots=None, mean=0, std=0.1, seed=N
...
-def cvqnn_layers_normal(n_layers, n_wires, mean=0, std=1, mean_active=0, std_active=0.1, seed=None):
-    r"""Creates a list of eleven parameter arrays for :func:`~.CVNeuralNetLayers`, where both active and non-active
-    gate parameters are drawn from normal distributions.
+        Keyword Args:
+            low (float): minimum value of uniform distribution
+            high (float): maximum value of uniform distribution
+            seed (int): seed used in sampling the parameters, makes function call deterministic
+
+        Returns:
+            parameter array
```
Comment (josh146) [inline pennylane/init.py]: just in general need to add return types

File: tests/test_init.py [hunk 7]
```diff
@@ -17,542 +17,194 @@
...
+        p_mean = np.mean(p)
+        assert np.isclose(p_mean, 1, atol=tol)
+
+    @pytest.mark.parametrize("init", NRML)
+    def test_interval_normal(self, init, seed, n_wires, tol):
+        """Test that normal samples of initialization functions lie
+        in correct interval."""
+        p = init(n_wires=n_wires, mean=1, std=0, seed=seed)
+        p_mean = np.mean(p)
+        assert np.isclose(p_mean, 1, atol=tol)
```
Comment (josh146) [inline tests/test_init.py]: Should we have integration tests in this file. That is, check that the generated parameters are accepted by the templates

File: tests/test_templates.py [hunk 8]
```diff
@@ -12,106 +12,256 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 """
-Unit tests for the integration of the methods from :mod:`pennylane.init` with the :mod:`pennylane.templates` module.
+Integration tests for templates, including integration with initialization functions
+in :mod:`pennylane.init`, running templates in larger circuits,
```
Comment (josh146) [inline tests/test_templates.py]: oh, integration tests are here 😆

File: tests/test_templates.py [hunk 9]
```diff
@@ -12,106 +12,256 @@
...
+combining templates, using positional and keyword arguments, and using different interfaces.
+
+New tests are added as follows:
+
+* When adding a new interface, try to import it and extend the fixture ``interfaces``.
+
+* When adding a new template, extend the fixtures ``qubit_const`` or ``cv_const`` by a *list* of arguments to the
+template. Note: Even if the template takes only one argument, it has to be wrapped in a list (i.e. [weights]).
```
Comment (josh146) [inline tests/test_templates.py]: Will this 'double nesting' work okay with TF?

File: tests/test_templates.py [hunk 10]
```diff
@@ -12,106 +12,256 @@
...
-        @qml.qnode(gaussian_device)
-        def circuit(weights):
-            CVNeuralNetLayers(*weights, wires=range(n_subsystems))
-            return qml.expval(qml.Identity(0))
+try:
+    import tensorflow as tf
 
-        circuit(weights=p)
+    if tf.__version__[0] == "1":
+        print(tf.__version__)
```
Comment (josh146) [inline tests/test_templates.py]:
```suggestion
```

File: .github/CHANGELOG.md [hunk 11]
```diff
@@ -11,8 +12,16 @@
 
 ### Breaking changes
 
+* Initialization functions now return single array of weights per function. Utility for multi-weight templates
+  `Interferometer()` and `CVNeuralNetLayers()` are provided.
+  [#XXX](https://github.com/XanaduAI/pennylane/pull/XXX)
```
Comment (josh146) [inline .github/CHANGELOG.md]:
```suggestion
  [#412](https://github.com/XanaduAI/pennylane/pull/412)
```

**Comment by mariaschuld** [comment]
This one is now ready to (re-)review. It contains - and partially overrides - changes from #409.

**Comment by mariaschuld** [comment]
Again ready for re-review...

File: tests/test_init.py [hunk 12]
```diff
@@ -17,542 +17,148 @@
...
+        ]
+# Functions returning a list of parameter arrays
+ALL = [(qml.init.cvqnn_layers_all, rpt),
+       (qml.init.interferometer_all, base)]
+
+###########
+# Shapes
+
+SHP = [(repeat, n_rots), (repeat, n_wires, 3), intf, intf, wrs, wrs,
+       wrs, wrs, wrs, wrs, (n_if,), (n_if,), (n_wires,),
```
Comment (co9olguy) [inline tests/test_init.py]: this one was hard to parse 😄

Comment (co9olguy) [reply tests/test_init.py]: Any way we can make this part clearer to people who might read and/or modify these tests?

File: tests/test_init.py [hunk 13]
```diff
@@ -17,542 +17,148 @@
...
+###############
+# Combinations
+
+SHAPES = [i + (s,) for i, s in zip(INIT, SHP)]
+ALL_SHAPES = [i + (s,) for i, s in zip(ALL, ALL_SHP)]
+
+##################
+
+
+class TestInitRepeated:
```
Comment (co9olguy) [inline tests/test_init.py]: I 😍 how much code this replaced.

File: tests/test_init.py (hunk 12 above)
Comment (co9olguy) [reply tests/test_init.py]: One suggestion: just using full words (i.e., `wires` instead of `wrs`) might make the cognitive load slightly easier

File: tests/test_init.py [hunk 14]
```diff
@@ -17,542 +17,148 @@
...
+    """Tests the initialization functions from the ``init`` module."""
+
+    @pytest.mark.parametrize("init, sgntr, shp", SHAPES)
+    def test_shape(self, init, sgntr, shp, seed):
+        """Confirm that initialization functions
+         return an array with the correct shape."""
+        s = sgntr.copy()
+        s['seed'] = seed
```
Comment (co9olguy) [inline tests/test_init.py]: perhaps instead of copying and updating the dict, we could just create a new dict which adds the seed?
Something like:
```
kwargs = {**s, 'seed':seed}
p = init(**kwargs)
```

Comment (co9olguy) [reply tests/test_init.py]: (applies throughout)

File: tests/test_init.py [hunk 15]
```diff
@@ -17,542 +17,148 @@
...
+        assert not np.allclose(p1, p2, atol=tol)
+
+    @pytest.mark.parametrize("init, sgntr", INIT)
+    def test_interval(self, init, sgntr, seed, tol):
+        """Test that sampled parameters lie in correct interval."""
+        s = sgntr.copy()
+        s['seed'] = seed
+
+        # Case A: Uniformly distributed parameters
+        if 'low' in sgntr.keys() and 'high' in s.keys():
```
Comment (co9olguy) [inline tests/test_init.py]: why check for `low` in `sgntr` but then `high` in `s`?

File: tests/test_init.py (hunk 13 above)
Comment (co9olguy) [reply tests/test_init.py]: I really like these new tests, they are compact and their intention is clear. 👍 
However, I think the parametrize arguments could be make a bit more readable, specifically `SHP`, `ALL_SHP`, `SHAPES`, and `ALL_SHAPES`

File: pennylane/init.py [hunk 16]
```diff
@@ -183,321 +131,643 @@ def random_layers_normal(n_layers, n_wires, n_rots=None, mean=0, std=0.1, seed=N
...
+        np.random.seed(seed)
+
+    varphi = np.random.uniform(low=low, high=high, size=(n_layers, n_wires))
+    return varphi
+
+
+def cvqnn_layers_varphi_normal(n_layers, n_wires, mean=0, std=0.1, seed=None):
+    r"""Creates a parameter array for the ``varphi`` input to the interferometers of :func:`~.CVNeuralNetLayers`.
+
+        The parameters are drawn from a uniform distribution.
```
Comment (co9olguy) [inline pennylane/init.py]:
```suggestion
        The parameters are drawn from a normal distribution.
```

File: pennylane/init.py [hunk 17]
```diff
@@ -183,321 +131,643 @@ def random_layers_normal(n_layers, n_wires, n_rots=None, mean=0, std=0.1, seed=N
...
-        std (float): standard deviation of non-active parameters
-        mean_active (float): mean of active gate parameters
-        std_active (float): standard deviation of active gate parameters
+        low (float): minimum value of uniform distribution
+        high (float): maximum value of uniform distribution
         seed (int): seed used in sampling the parameters, makes function call deterministic
 
     Returns:
-        list of parameter arrays
+        list: list of parameter arrays
```
Comment (co9olguy) [inline pennylane/init.py]: How come changed here but not in `interferometer_all` or `interferometer_theta_uniform`?

Comment (co9olguy) [reply pennylane/init.py]: In fact, the docstrings of these functions each say they create `a parameter array`. @mariaschuld can you double check?

File: pennylane/init.py [hunk 18]
```diff
@@ -183,321 +131,643 @@ def random_layers_normal(n_layers, n_wires, n_rots=None, mean=0, std=0.1, seed=N
...
     Keyword Args:
-        low (float): minimum value of uniformly drawn rotation angles
-        high (float): maximum value of uniformly drawn rotation angles
+        mean(float): mean of normal distribution
+        std(float): standard deviation of normal distribution
         seed (int): seed used in sampling the parameters, makes function call deterministic
 
     Returns:
-        list of parameter arrays
+        list: list of parameter arrays
```
Comment (co9olguy) [inline pennylane/init.py]: see above comment

File: pennylane/init.py [hunk 19]
```diff
@@ -183,321 +131,643 @@ def random_layers_normal(n_layers, n_wires, n_rots=None, mean=0, std=0.1, seed=N
...
+
+        Keyword Args:
+            low (float): minimum value of uniform distribution
+            high (float): maximum value of uniform distribution
+            seed (int): seed used in sampling the parameters, makes function call deterministic
+
+        Returns:
+            array: parameter array
+        """
     if seed is not None:
```
Comment (co9olguy) [inline pennylane/init.py]: *almost* feels like we should make a simple helper function do avoid a lot of duplicate code. Something like
```
def _sample_uniform(low, high, size, seed=None):
    if seed is not None:
        np.random.seed(seed)
    return np.random.uniform(low=low, high=high, size=size)
```

Comment (co9olguy) [reply pennylane/init.py]: (more maintainable too, since absolutely every `init` function would be fed through the same 2 helper functions)

**Comment by mariaschuld** [comment]
Thanks @co9olguy ! I addressed all your comments, except from the one avoiding code repetition which I added to the issue on further template test improvements. 

Ready for re-review.

Note: Some of your comments were related to changes PR#409. I don't mind, but just don't get confused if they are updated on this branch but not in the original PR.

File: .github/CHANGELOG.md [hunk 20]
```diff
@@ -69,10 +73,6 @@
 
 This release contains contributions from (in alphabetical order):
 
-<<<<<<< HEAD
-
-=======
->>>>>>> master
 Ville Bergholm, Josh Izaac, Angus Lowe, Johannes Jakob Meyer, Oluwatobi Ogunbayo, Maria Schuld
```
Comment (mariaschuld) [inline .github/CHANGELOG.md]: Not sure what this is doing on master

File: doc/introduction/templates.rst [hunk 21]
```diff
@@ -228,6 +228,7 @@ The following initialization functions are available:
 .. autosummary::
     :nosignatures:
 
+<<<<<<< HEAD
```
Comment (co9olguy) [inline doc/introduction/templates.rst]: more merge issues here

File: tests/test_init.py [hunk 22]
```diff
@@ -71,47 +71,45 @@
...
 # Shapes
 
-SHP = [(repeat, n_rots), (repeat, n_wires, 3), intf, intf, wrs, wrs,
-       wrs, wrs, wrs, wrs, (n_if,), (n_if,), (n_wires,),
-       (repeat, n_rots), (repeat, n_wires, 3), intf, intf, wrs, wrs,
-       wrs, wrs, wrs, wrs, (n_if,), (n_if,), (n_wires,),
-       ]
-ALL_SHP = [[intf, intf, wrs, wrs, wrs, intf, intf, wrs, wrs, wrs, wrs],
-           [(n_if,), (n_if,), (n_wires,)]]
+SHAPE = [(repeat, n_rots), (repeat, n_wires, 3), intf_block, intf_block, wire_block, wire_block,
```
Comment (co9olguy) [inline tests/test_init.py]: The names are better, but I still find myself struggling to parse easily what these elements refer to.
Suggestion: Can we recast it so the shapes are right next to the corresponding inits? (essentially a hand-written `zip`)
```
INIT_AND_SHAPES = [[(qml.init.random_layers_normal, rnd_rpt_nrml), (repeat, n_rots)], 
                   [(qml.init.strong_ent_layers_normal, rpt_nrml), (repeat, n_wires, 3)],
                   ...
                  ]
```
From that, we could easily make separate `INIT` and `SHAPE` objects programmatically using list comprehensions

File: doc/introduction/templates.rst (hunk 21 above)
Comment (mariaschuld) [reply doc/introduction/templates.rst]: Very sorry, removed and made sure that there are no merge-like things in the entire path...

File: tests/test_templates.py [hunk 23]
```diff
@@ -83,13 +86,15 @@
...
+# qubit templates and constant inputs
 qubit_const = [(StronglyEntanglingLayers, [[[[4.54, 4.79, 2.98], [4.93, 4.11, 5.58]],
                                            [[6.08, 5.94, 0.05], [2.44, 5.07, 0.95]]]]),
                (RandomLayers, [[[0.56, 5.14], [2.21, 4.27]]]),
-               (AmplitudeEmbedding, [[1 / 2, 1 / 2, 1 / 2, 1 / 2]]),
-               (BasisEmbedding, [[1, 0]]),
-               (AngleEmbedding, [[1., 2.]])
-              ]
+               (AngleEmbedding, [[1., 2.]]),
+               ]
```
[the remark this answers is not in the document]
Comment (mariaschuld) [reply tests/test_templates.py]: Yes! Johannes PR changed this.

File: tests/test_init.py (hunk 22 above)
Comment (mariaschuld) [reply tests/test_init.py]: I will address this in the PR #419, since I changed the signature slightly there. Want to avoid merge conflicts.

File: tests/test_templates.py [hunk 24]
```diff
@@ -18,13 +18,14 @@
 
 New tests are added as follows:
 
-* When adding a new interface, try to import it and extend the fixture ``interfaces``.
+* When adding a new interface, try to import it and extend the fixture ``interfaces``. Add interface to
+  TestGradientIntegration tests.
```
[the remark this answers is not in the document]
Comment (mariaschuld) [reply tests/test_templates.py]: I clarified this statement.

**Comment by mariaschuld** [comment]
@josh146 , any changes requested from your side?

**Comment by mariaschuld** [comment]
@co9olguy I addressed all your requests.

File: doc/introduction/templates.rst [hunk 25]
```diff
@@ -152,8 +228,24 @@ randomly initialized arrays for the trainable parameters.
...
+    ~pennylane.init.cvqnn_layers_r_uniform
+    ~pennylane.init.cvqnn_layers_r_normal
+    ~pennylane.init.cvqnn_layers_phi_r_uniform
+    ~pennylane.init.cvqnn_layers_phi_r_normal
+    ~pennylane.init.cvqnn_layers_a_uniform
+    ~pennylane.init.cvqnn_layers_a_normal
+    ~pennylane.init.cvqnn_layers_phi_a_uniform
+    ~pennylane.init.cvqnn_layers_phi_a_normal
+    ~pennylane.init.cvqnn_layers_kappa_uniform
+    ~pennylane.init.cvqnn_layers_kappa_normal
```
Comment (josh146) [inline doc/introduction/templates.rst]: Instead of listing all these hear, can we just link to the `qml.init` api page, where they are all listed as well? Avoids duplication and saves with maintenance.

File: doc/introduction/templates.rst [hunk 26]
```diff
@@ -164,7 +256,12 @@ randomly initialized arrays for the trainable parameters.
...
 
-    ~pennylane.init.interferometer_uniform
-    ~pennylane.init.interferometer_normal
+    ~pennylane.init.interferometer_all
+    ~pennylane.init.interferometer_theta_uniform
+    ~pennylane.init.interferometer_theta_normal
+    ~pennylane.init.interferometer_phi_uniform
+    ~pennylane.init.interferometer_phi_normal
+    ~pennylane.init.interferometer_varphi_uniform
+    ~pennylane.init.interferometer_varphi_normal
```
Comment (josh146) [inline doc/introduction/templates.rst]: Ditto here.

**Comment by co9olguy** [comment]
> @co9olguy I addressed all your requests.

Yep already approved from my side

File: doc/introduction/templates.rst (hunk 25 above)
Comment (mariaschuld) [reply doc/introduction/templates.rst]: I propose we do this in a separate docs PR that adapts the docs to the templates redesign.
